src/Storages/ObjectStorage/StorageObjectStorageSource.h

```diff
@@ -0,0 +1,221 @@
+#pragma once
+#include <Processors/SourceWithKeyCondition.h>
+#include <Processors/Executors/PullingPipelineExecutor.h>
+#include <Interpreters/Context_fwd.h>
+#include <Storages/ObjectStorage/StorageObjectStorage.h>
+#include <Processors/Formats/IInputFormat.h>
+#include <Common/re2.h>
+
+
+namespace DB
+{
+
+class SchemaCache;
+
+class StorageObjectStorageSource : public SourceWithKeyCondition, WithContext
+{
+    friend class StorageS3QueueSource;
+public:
+    using ConfigurationPtr = StorageObjectStorage::ConfigurationPtr;
+    using ObjectInfo = StorageObjectStorage::ObjectInfo;
+    using ObjectInfos = StorageObjectStorage::ObjectInfos;
+    using ObjectInfoPtr = StorageObjectStorage::ObjectInfoPtr;
+
+    class IIterator;
+    class ReadTaskIterator;
+    class GlobIterator;
+    class KeysIterator;
+
+    StorageObjectStorageSource(
+        String name_,
+        ObjectStoragePtr object_storage_,
+        ConfigurationPtr configuration,
+        const ReadFromFormatInfo & info,
+        std::optional<FormatSettings> format_settings_,
+        ContextPtr context_,
+        UInt64 max_block_size_,
+        std::shared_ptr<IIterator> file_iterator_,
+        bool need_only_count_);
+
+    ~StorageObjectStorageSource() override;
+
+    String getName() const override { return name; }
+
+    void setKeyCondition(const ActionsDAGPtr & filter_actions_dag, ContextPtr context_) override;
+
+    Chunk generate() override;
+
+    static std::shared_ptr<IIterator> createFileIterator(
+        ConfigurationPtr configuration,
+        ObjectStoragePtr object_storage,
+        bool distributed_processing,
+        const ContextPtr & local_context,
+        const ActionsDAG::Node * predicate,
+        const NamesAndTypesList & virtual_columns,
+        ObjectInfos * read_keys,
+        std::function<void(FileProgress)> file_progress_callback = {});
+
+protected:
+    const String name;
+    ObjectStoragePtr object_storage;
+    const ConfigurationPtr configuration;
+    const std::optional<FormatSettings> format_settings;
+    const UInt64 max_block_size;
+    const bool need_only_count;
+    const ReadFromFormatInfo read_from_format_info;
```

Comment:
I am wondering if we can store `const ReadFromFormatInfo &` here to avoid copying it to each source in `ReadFromObjectStorageStep::initializePipeline`. From the first glance, `ReadFromObjectStorageStep` should live longer than `StorageObjectStorageSource` but I am not sure (maybe we can ask Nikolay)